File: app/src/main/java/com/mycompany/myapp/ui/main/MainActivity.java
package com.mycompany.myapp.ui.main;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.design.widget.Snackbar;

import com.mycompany.myapp.MainActivityBinding;
import com.mycompany.myapp.R;
import com.mycompany.myapp.app.MainApplication;
import com.mycompany.myapp.ui.BaseActivity;
import com.mycompany.myapp.ui.main.MainComponent.MainModule;
import com.mycompany.myapp.ui.main.MainFragment.MainFragmentHost;
import com.mycompany.myapp.ui.main.MainPresenter.MainViewContract;

import javax.inject.Inject;

public class MainActivity extends BaseActivity implements MainViewContract, MainFragmentHost {
    private MainComponent component;

    @Inject MainPresenter presenter;

    private MainActivityBinding binding;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        component = ((MainApplication) getApplication()).getComponent()
                .mainComponent(new MainModule(this));
        component.inject(this);

        presenter.setView(this);
        lifecycleRegistry.addObserver(presenter);
        presenter.restoreState(savedInstanceState);

        super.onCreate(savedInstanceState);

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setPresenter(presenter);

        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setTitle(getApplicationContext().getPackageName());
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        presenter.saveState(outState);
    }

    @Override
    public void inject(MainFragment fragment) {
        component.inject(fragment);
    }

    @Override
    public void displayError(String message) {
        Snackbar.make(binding.rootLayout, message, Snackbar.LENGTH_LONG).show();
    }
}

File: app/src/main/java/com/mycompany/myapp/ui/main/MainActivity.kt
package com.mycompany.myapp.ui.main

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.Snackbar
import com.mycompany.myapp.R
import com.mycompany.myapp.ui.BaseActivity
import com.mycompany.myapp.ui.SimpleSnackbarMessage
import com.mycompany.myapp.ui.main.MainFragment.MainFragmentHost

class MainActivity : BaseActivity(), MainFragmentHost {
    private lateinit var viewModel: MainViewModel
    private lateinit var binding: MainActivityBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        appComponent.inject(this)
        super.onCreate(savedInstanceState)

        viewModel = getViewModel(MainViewModel::class)
        viewModel.restoreState(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        binding.vm = viewModel
        binding.executePendingBindings()

        setSupportActionBar(binding.toolbar)
        supportActionBar!!.title = applicationContext.packageName

        viewModel.snackbarMessage.observe(this, object : SimpleSnackbarMessage.SnackbarObserver {
            override fun onNewMessage(message: String) {
                Snackbar.make(binding.rootLayout, message, Snackbar.LENGTH_LONG).show()
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        viewModel.saveState(outState)
    }

    override fun onResume() {
        super.onResume()
        viewModel.onResume()
    }
}